packages/protocol/contracts/common/libraries/CIP20Lib.sol

```diff
@@ -0,0 +1,141 @@
+// From https://github.com/prestwich/cip20-sol
+pragma solidity ^0.5.13;
+
+library CIP20Lib {
+  uint8 private constant CIP20_ADDRESS = 0xE2;
+
+  uint8 private constant SHA3_256_SELECTOR = 0x00;
+  uint8 private constant SHA3_512_SELECTOR = 0x01;
+  uint8 private constant KECCAK_512_SELECTOR = 0x02;
+  uint8 private constant SHA2_512_SELECTOR = 0x03;
+  uint8 private constant BLAKE2S_SELECTOR = 0x10;
+
+  // BLAKE2S_DEFAULT_CONFIG = createBlake2sConfig(32, 0, 1, 1, 0, 0, 0, 0, 0, 0, 0)
+  bytes32 private constant BLAKE2S_DEFAULT_CONFIG = 0x2000010100000000000000000000000000000000000000000000000000000000;
+
+  // Accepts a fully formed input blob. This should include any config
+  // options and the preimage, but not the selector.
+  function executeCIP20(bytes memory input, uint8 selector, uint256 output_len)
+    internal
+    view
+    returns (bytes memory)
+  {
+    uint8 addr = CIP20_ADDRESS;
+    bytes memory output = new bytes(output_len);
+
+    // To avoid copying the input array (an unbounded cost) we store its
+    // length on the stack and then replace the length prefix for its
+    // in-memory representation with the selector. We then replace the
+    // length in memory after the precompile executes with it.
+    uint256 len = input.length;
+
+    bool success;
+    assembly {
+      mstore(input, selector) // selector
+
+      success := staticcall(
+        sub(gas, 2000),
+        addr,
+        add(input, 0x1F), // location is shifted 1 byte for selector
+        add(len, 0x01), // length w/ selector
+        add(output, 0x20), // location
+        mload(output) // length
+      )
+
+      // Restore the input array length prefix
+      mstore(input, len)
+    }
+
+    require(success, "CIP-20 hashing failed");
+    return output;
+  }
+
+  function sha3_256(bytes memory input) internal view returns (bytes memory) {
+    return executeCIP20(input, SHA3_256_SELECTOR, 32);
+  }
+
+  function sha3_512(bytes memory input) internal view returns (bytes memory) {
+    return executeCIP20(input, SHA3_512_SELECTOR, 64);
+  }
+
+  function keccak512(bytes memory input) internal view returns (bytes memory) {
+    return executeCIP20(input, KECCAK_512_SELECTOR, 64);
+  }
+
+  function sha2_512(bytes memory input) internal view returns (bytes memory) {
+    return executeCIP20(input, SHA2_512_SELECTOR, 64);
+  }
+
+  function blake2sWithConfig(bytes32 config, bytes memory key, bytes memory preimage)
+    internal
+    view
+    returns (bytes memory)
+  {
+    require(
+      key.length == uint256(config >> (8 * 30)) & 0xff,
+      "CIP20Lib/blake2sWithConfig - Provided key length does not match key length in config"
+    );
+    bytes memory configuredInput = abi.encodePacked(config, preimage);
+    return executeCIP20(configuredInput, BLAKE2S_SELECTOR, uint256(uint8(config[0])));
+  }
+
+  // default settings, no key
```

Comment:
same. The comments didn't really help me - perharps we could had some more explanation of what you mean, otherwise we can check if it is actually needed. Great work btw

Reply:
or we can rename function to `blake2sWithoutConfig` 
seems verbose howerver the entire code could help with further handholding cos i'm not too familiar with whats going on mostly anyways 🤷